src/librustc_mir/transform/const_prop.rs

```diff
@@ -539,13 +560,28 @@ impl<'a, 'mir, 'tcx> ConstPropagator<'a, 'mir, 'tcx> {
                         *rval = Rvalue::Aggregate(
                             Box::new(AggregateKind::Tuple),
                             vec![
-                                self.operand_from_scalar(one, substs[0].expect_ty(), span),
-                                self.operand_from_scalar(two, substs[1].expect_ty(), span),
+                                self.operand_from_scalar(
+                                    one, substs[0].expect_ty(), source_info.span
+                                ),
+                                self.operand_from_scalar(
+                                    two, substs[1].expect_ty(), source_info.span
+                                ),
                             ],
                         );
                     }
                 },
                 _ => { }
+
+            },
+            interpret::Operand::Indirect(MemPlace{ ptr, .. }) => {
+                let op = self.operand_from_scalar(ptr, value.layout.ty, source_info.span);
+                if let Some(opty) = self.eval_operand(&op, source_info) {
+                    if let Ok(mplace) = self.ecx.deref_operand(opty) {
```

Comment:
The above three lines seem like something we'd have abstractions for that work directly on `interpret::Operand`. I believe https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/interpret/struct.InterpretCx.html#method.read_immediate should allow you to turn an `OpTy` into an `Immediate` that you can then match on. So instead of matching on `*value` above, call `read_immedidate` and then just have the match on `im` (returned by `read_immediate`).

I may be misunderstanding what is happening here, so take the above with a grain of salt.